src/server/scripts/Kalimdor/Firelands/boss_ragnaros_firelands.cpp

```diff
@@ -0,0 +1,5304 @@
+    { 998.0f,  -103.0f, 55.5f, 0.0f },
+    { 998.0f,  -13.0f,  55.5f, 0.0f },
+    { 998.0f,  -18.0f,  55.5f, 0.0f },
+    { 998.0f,  -23.0f,  55.5f, 0.0f },
+    { 998.0f,  -28.0f,  55.5f, 0.0f },
+    { 998.0f,  -33.0f,  55.5f, 0.0f },
+    { 998.0f,  -38.0f,  55.5f, 0.0f },
+    { 998.0f,  -43.0f,  55.5f, 0.0f },
+    { 998.0f,  -48.0f,  55.5f, 0.0f },
+    { 998.0f,  -53.0f,  55.5f, 0.0f },
+    { 998.0f,  -58.0f,  55.5f, 0.0f },
+    { 998.0f,  -63.0f,  55.5f, 0.0f },
+    { 998.0f,  -68.0f,  55.5f, 0.0f },
+    { 998.0f,  -73.0f,  55.5f, 0.0f },
+    { 998.0f,  -78.0f,  55.5f, 0.0f },
+    { 998.0f,  -83.0f,  55.5f, 0.0f },
+    { 998.0f,  -88.0f,  55.5f, 0.0f },
+    { 998.0f,  -93.0f,  55.5f, 0.0f },
+    { 998.0f,  -98.0f,  55.5f, 0.0f }
+};
+
+class RagnarosAuraRemovePredicate
+{
+    public:
+        RagnarosAuraRemovePredicate(WorldObject const* refObj) : _refObj(refObj) { }
+
+        bool operator()(uint32 spellId) const
+        {
+            switch (spellId)
+            {
+                case SPELL_BASE_VISUAL:
+                case SPELL_BURNING_WOUNDS:
+                    return false;
+                default:
+                    return true;
+            }
+        }
+
+    private:
+        WorldObject const* _refObj;
+};
+
+class HeartOfFlameCheck : public std::unary_function<WorldObject*, bool>
+{
+    public:
+        HeartOfFlameCheck() : _questId(0) { }
+        HeartOfFlameCheck(uint32 questId) : _questId(questId) { }
+
+        bool operator()(WorldObject* target) const
+        {
+            Player* player = target->ToPlayer();
+            if (!player)
+                return true;
+
+            // Get faction specific legendary quest and check if it's in progress
+            Quest const* quest = sObjectMgr->GetQuestTemplate(_questId != 0 ? _questId : player->GetTeam() == ALLIANCE ? QUEST_HEART_OF_FLAME_A : player->GetTeam() == HORDE ? QUEST_HEART_OF_FLAME_H : 0);
+            if (!quest || player->GetQuestStatus(quest->ID) != QUEST_STATUS_INCOMPLETE)
+                return true;
+
+            // Ensure the quest criterias we're looking for are met, in this case player should have collected all the smouldering essences and does not possess the heart of flame already
+            for (QuestObjective const& obj : quest->GetObjectives())
+                if (obj.ObjectID == ITEM_HEART_OF_FLAME && player->HasItemCount(ITEM_HEART_OF_FLAME, obj.Amount) ||
+                    obj.ObjectID == NPC_SMOULDERING_ESSENCE_CREDIT && player->GetReqKillOrCastCurrentCount(quest->ID, NPC_SMOULDERING_ESSENCE_CREDIT) != obj.Amount)
+                    return true;
+
+            return false;
+        }
+
+    private:
+        uint32 _questId;
+};
+
+class DelayedDisappearAndDieEvent : public BasicEvent
+{
+    public:
+        DelayedDisappearAndDieEvent(Creature* owner) : _owner(owner) { }
+
+        bool Execute(uint64 /*execTime*/, uint32 /*diff*/) override
+        {
+            _owner->DisappearAndDie();
```

Comment:
Hmm... I wonder why I used this. DespawnOrUnsummon comes to mind as the obvious function to use for this behavior, but I wonder if there was an underlying reason, unwanted behavior perhaps, or other side-effect.
I'll have to check on this.